### ISSUE TITLE
Allow passing an entire object to AutoCompleteItem as the value

### ISSUE THREAD

**Issue by jamesopti** [anubra266/choc-autocomplete#60]
# Overview

It's very common to use a component like this with a list of objects. While the input value is always a string, it would be nice to also pass the entire object to `AutoComplete` and invoke the `onSelectOption` with that value.

## Example

```ts
<AutoComplete
  // This is a hack to reset the internal state of the component
  key={instanceCount}
  ...
  onSelectOption={(selection) => {
    // selection is { email: 'you@me.com', id: 's3e8d' }
  }}
>
  ...
  <AutoCompleteItem
    key={item.email}
    value={{ email: 'you@me.com', id: 's3e8d' }}
  />
```

**Comment by anubra266** [comment]
The list is filteredd based on the string value, how do you suggest this fits in.

**Comment by jamesopti** [comment]
> The list is filteredd based on the string value, how do you suggest this fits in.

Isn't the list [filtered by the caller (which is optional)](https://github.com/anubra266/choc-autocomplete/blob/main/src/use-autocomplete.ts#L92-L104)? I'm hooking it up to a search API, so I expect the full list to be whatever I pass in.

While changing the value input to an object would be a larger refactor, it would align closer to the MaterialUI Autocomplete API: https://material-ui.com/api/autocomplete/

I could see 2 options that are more incremental though:

1. Provide `item` as an optional prop and have it passed to `onSelectionUpdate` along with `selection` as a convenience.
  ```ts
    onSelectOption={(selection, item) => {
      // selection is 'you@me.com'
      // item is { email: 'you@me.com', id: 's3e8d' }
    }}
    ...
    <AutoCompleteItem
      key={item.email}
      value={item.email}
      item={{ email: 'you@me.com', id: 's3e8d' }}
  />
  ```


2. Enable a `getValue` method to be passed that indicates how to lookup the string. If `typeof value === 'object'`, then `getValue` is used (defaults to `(item) => item.value`)

  ```ts
    <AutoCompleteItem
      key={item.email}
      getValue={ (item) => item.email }
      value={{ email: 'you@me.com', id: 's3e8d' }}
  />
  ```

**Comment by jamesopti** [comment]
For more context, my current workaround is to find the user by email address on selection, which is brittle and makes it even harder to detect the "Creatable" user

```ts
onSelectOption={(selection) => {
  const email = selection.optionValue
  const existingUser = searchUsers.find((u) => u.email === email)
  setSelectedUser(existingUser ? existingUser : { email })
}}
```

**Comment by anubra266** [comment]
Okay @jamesopti I'll go with the first option, and add an `item` prop.

**Comment by anubra266** [comment]
Hi @jamesopti, really sorry it took the long. Just finished my exams. The `value` prop now accepts objects, and the `getValue` prop is now available to resolve the value used by default. If the `getValue`  prop does not exist and you pass an object as the value, the first key in the value is used. See changes in [v4.8.0](https://github.com/anubra266/choc-autocomplete/releases/tag/v4.8.0)

**Comment by SpencerKaiser** [comment]
@anubra266 what's the expectation if we want the `value` to just be an object?

I think I have a use case similar to @jamesopti where each option represents a pair of data but I don't want to display half of it to the user (an `address` and the corresponding `placeId` from Google Maps). I thought maybe passing the pair as an object into the `value` would be a good workaround, but after reading the above, I'm not sure that's what I should be doing.

Depending on whether you have a good suggestion for how use an option, I think my next-best approach is to just do a lookup for the address's corresponding placeId on selection.

**Comment by anubra266** [comment]
@SpencerKaiser That's what this allows, just pass the object as the value of your Item, it can then be received in the `onChange` prop.

**Comment by SpencerKaiser** [comment]
So would I need to modify `getValue` to just return the full object? I read through the docs and it looks like you default to pulling out the value associated with the first key, which is what happened, but I was worried about trying the full object for some reason 😅 

Just tried it now and it looks like I'm getting an error because `toLowerCase` isn't a method of the object I passed in as the value.


``` typescript
interface AddressRecommendation {
  address: string;
  placeId: string;
}

// address below is an `AddressRecommendation`
<AutoCompleteItem key={idx} value={address} getValue={(address) => address}>
  {address.address}
</AutoCompleteItem>
```

```
Unhandled Runtime Error
TypeError: itemValue.toLowerCase is not a function. (In 'itemValue.toLowerCase()', 'itemValue.toLowerCase' is undefined)
```

I ended up refactoring it to where `onSelection` just looks up the corresponding `placeId` for the selected address, so it's not a big deal, but figured I'd add extra context to help clarify what I saw as a limitation (and I think @jamesopti was on the same page).

**Comment by anubra266** [comment]
@SpencerKaiser `getValue` tells what should be used in searching the item, `getValue={(address) => address.address}`

**Comment by SpencerKaiser** [comment]
Alright I just got it all setup again and this was the issue... 

``` typescript
<AutoCompleteItem
  key={idx}
  value={address}
  getValue={(address: AddressRecommendation) => address.address}
>
  {address.address}
</AutoCompleteItem>
```
The above works as expected, but when I select an item and set a debugger within `onSelection`, here's what `selectedOption` is:
![image](https://user-images.githubusercontent.com/6445731/151414074-8d62f36c-d142-4efb-a1e1-ffe727a4c444.png)

For my use case, I would expect `selectedOption.item.value` to be the full object so I could set the selected `address` and `placeId`, not just the value of the key I provided in `getValue`. Make sense? I wouldn't make this change without getting input from others, especially since it's a breaking change, but hopefully that clears up what my intent was.
